### PR TITLE
[Spark] Managed Commits: add a DynamoDB-based commit owner

### PR DESCRIPTION
**Pull request by dhruvarya-db**
#### Which Delta project/connector is this regarding?

- [X] Spark
- [ ] Standalone
- [ ] Flink
- [ ] Kernel
- [ ] Other (fill in here)

## Description

Taking inspiration from https://github.com/delta-io/delta/pull/339, this PR adds a Commit Owner Client which uses DynamoDB as the backend. Each Delta table managed by a DynamoDB instance will have one corresponding entry in a DynamoDB table. The table schema is as follows:

* tableId: String --- The unique identifier for the entry. This is a UUID.
* path: String --- The fully qualified path of the table in the file system. e.g. s3://bucket/path.
* acceptingCommits: Boolean --- Whether the commit owner is accepting new commits. This will only
*  be set to false when the table is converted from managed commits to file system commits.
* tableVersion: Number --- The version of the latest commit.
* tableTimestamp: Number --- The inCommitTimestamp of the latest commit.
* schemaVersion: Number --- The version of the schema used to store the data.
* commits: --- The list of unbackfilled commits.
  -  version: Number --- The version of the commit.
  -  inCommitTimestamp: Number --- The inCommitTimestamp of the commit.
  -  fsName: String --- The name of the unbackfilled file.
  -  fsLength: Number --- The length of the unbackfilled file.
  -  fsTimestamp: Number --- The modification time of the unbackfilled file.

For a table to be managed by DynamoDB, `registerTable` must be called for that Delta table. This will create a new entry in the db for this Delta table. Every `commit` invocation appends the UUID delta file status to the `commits` list in the table entry. `commit` is performed through a conditional write in DynamoDB.

## How was this patch tested?

Added a new suite called `DynamoDBCommitOwnerClient5BackfillSuite` which uses a mock DynamoDB client. + plus manual testing against a DynamoDB instance.

### CODE REVIEW

File: spark/src/test/scala/org/apache/spark/sql/delta/managedcommit/CommitOwnerClientImplSuiteBase.scala [hunk 1]
```diff
@@ -198,15 +198,19 @@ trait CommitOwnerClientImplSuiteBase extends QueryTest
...
 
       val e = intercept[CommitFailedException] {
         commit(version = 0, timestamp = 0, tableCommitOwnerClient)
       }
       assert(e.getMessage === "Commit version 0 must go via filesystem.")
       writeCommitZero(logPath)
-      assert(tableCommitOwnerClient.getCommits() == GetCommitsResponse(Seq.empty, -1))
+      var expectedVersion = -1
+      if (tableCommitOwnerClient.commitOwnerClient.isInstanceOf[DynamoDBCommitOwnerClient]) {
+        // DynamoDBCommitOwnerClient stores attemptVersion as the current table version when
```
Comment (prakharjain09) [inline spark/src/test/scala/org/apache/spark/sql/delta/managedcommit/CommitOwnerClientImplSuiteBase.scala:209]: `DynamoDBCommitOwnerClient.getCommits` starts advertising `attemptVersion` as the current commit version even without knowledge around whether the commit has gone through or not. Practically this won't cause any issue because until the commit is done, no client will invoke `DynamoDBCommitOwnerClient.getCommits` since the table was a FS table. But just for the sake of semantics, we should fix it in a followup PR.

File: spark/src/main/java/io/delta/dynamodbcommitstore/DynamoDBCommitOwnerClient.java [hunk 2]
```diff
@@ -0,0 +1,681 @@
...
+            long backfillBatchSize) throws IOException {
+        this(
+            managedCommitsTableName,
+            endpoint,
+            client,
+            backfillBatchSize,
+            5 /* readCapacityUnits */,
+            5 /* writeCapacityUnits */,
+            false /* skipPathCheck */);
+        tryEnsureTableExists();
```
Comment (prakharjain09) [inline spark/src/main/java/io/delta/dynamodbcommitstore/DynamoDBCommitOwnerClient.java:130]: We are calling the other constructor here so it will take care of invoking tryEnsureTableExists ?

File: spark/src/main/java/io/delta/dynamodbcommitstore/DynamoDBCommitOwnerClient.java [hunk 3]
```diff
@@ -0,0 +1,681 @@
...
+                                "attribute_not_exists(%s)", DynamoDBTableEntryConstants.TABLE_ID));
+        client.putItem(request);
+
+        Tuple2<String, String> tableIdEntry = new Tuple2<>(DynamoDBTableEntryConstants.TABLE_ID, tableId);
+        Map<String, String> tableConf = Map$.MODULE$.<String, String>empty().$plus(tableIdEntry);
+
+        return tableConf;
+    }
+
+    // Copied from DynamoDbLogStore. TODO: add the logging back.
```
Comment (prakharjain09) [inline spark/src/main/java/io/delta/dynamodbcommitstore/DynamoDBCommitOwnerClient.java:618]: What logging needs to be added back?

File: spark/src/main/java/io/delta/dynamodbcommitstore/DynamoDBCommitOwnerClientBuilder.java [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+    // TODO: update this interface so that it can take a sparkSession.
+    @Override
+    public CommitOwnerClient build(Map<String, String> conf) {
+        String managedCommitsTableName = conf.get(MANAGED_COMMITS_TABLE_NAME_KEY).getOrElse(() -> {
+            throw new RuntimeException(MANAGED_COMMITS_TABLE_NAME_KEY + " not found");
+        });
+        String dynamoDBEndpoint = conf.get(DYNAMO_DB_ENDPOINT_KEY).getOrElse(() -> {
+            throw new RuntimeException(DYNAMO_DB_ENDPOINT_KEY + " not found");
+        });
+        String awsCredentialsProviderName = conf.get(AWS_CREDENTIALS_PROVIDER_KEY).getOrElse( () ->
```
Comment (prakharjain09) [inline spark/src/main/java/io/delta/dynamodbcommitstore/DynamoDBCommitOwnerClientBuilder.java:50]:
```suggestion
        String awsCredentialsProviderName = conf.get(AWS_CREDENTIALS_PROVIDER_KEY).getOrElse(() ->
```

Comment (prakharjain09) [inline spark/src/main/java/io/delta/dynamodbcommitstore/DynamoDBCommitOwnerClientBuilder.java:50]: Should AWS_CREDENTIALS_PROVIDER_KEY come from the commitOwnerConf from DeltaLog? Isn't this more of an authentication config which should be configured at JVM level i.e. via SparkSession.

File: spark/src/main/java/io/delta/dynamodbcommitstore/DynamoDBCommitOwnerClientBuilder.java [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+    private final long BACKFILL_BATCH_SIZE = 1L;
+
+    @Override
+    public String getName() {
+        return "dynamodb";
+    }
+
+    private static final String MANAGED_COMMITS_TABLE_NAME_KEY = "managedCommitsTableName";
+    private static final String DYNAMO_DB_ENDPOINT_KEY = "dynamoDBEndpoint";
+    private static final String AWS_CREDENTIALS_PROVIDER_KEY = "awsCredentialsProvider";
```
Comment (prakharjain09) [inline spark/src/main/java/io/delta/dynamodbcommitstore/DynamoDBCommitOwnerClientBuilder.java:39]: Add comments for these properties.

File: spark/src/main/java/io/delta/dynamodbcommitstore/DynamoDBCommitOwnerClient.java (hunk 2 above)
Comment (dhruvarya-db) [reply spark/src/main/java/io/delta/dynamodbcommitstore/DynamoDBCommitOwnerClient.java:130]: You are right, will fix

File: spark/src/main/java/io/delta/dynamodbcommitstore/DynamoDBCommitOwnerClient.java (hunk 3 above)
Comment (dhruvarya-db) [reply spark/src/main/java/io/delta/dynamodbcommitstore/DynamoDBCommitOwnerClient.java:618]: 1. That the table exists/does not exist
2. Table creation attempt
3. Table creation failure
I will add these and logs in other parts of this class in a follow up PR.

File: spark/src/main/java/io/delta/dynamodbcommitstore/DynamoDBCommitOwnerClientBuilder.java (hunk 4 above)
Comment (dhruvarya-db) [reply spark/src/main/java/io/delta/dynamodbcommitstore/DynamoDBCommitOwnerClientBuilder.java:50]: Yes, once the changes around passing sparkSession to the CommitOwnerBuilder have baked in, I will update this to read it directly from sparkSession. For now, I will hardcode this to the default credentials provider.

File: spark/src/test/scala/org/apache/spark/sql/delta/managedcommit/CommitOwnerClientImplSuiteBase.scala (hunk 1 above)
Comment (dhruvarya-db) [reply spark/src/test/scala/org/apache/spark/sql/delta/managedcommit/CommitOwnerClientImplSuiteBase.scala:209]: Yes, for DynamoDB, we treat registerTable as a valid commit and store the attemptVersion instead of the current Delta table version. This is not risky because the DynamoDB table entry is only treated as valid if the corresponding filesystem commit goes through (which means that the attemptVersion is in fact the Delta table version). In case of a filesystem conflict, it is expected that registerTable will be called again. However, for consistency, we can use another bool to track whether any commits have gone through DynamoDB and return -1.

File: spark/src/main/java/io/delta/dynamodbcommitstore/DynamoDBCommitOwnerClient.java [hunk 6]
```diff
@@ -0,0 +1,681 @@
...
+    }
+
+    /**
+     * Writes the given actions to a file.
+     * logStore.write(overwrite=false) will throw a FileAlreadyExistsException if the file already
+     * exists. However, the scala LogStore interface does not declare this as part of the function
+     * signature. This method wraps the write method and declares the exception to ensure that the
+     * caller is aware of the exception.
+     */
+    private void writeActionsToFile(
```
Comment (LukasRupprecht) [inline spark/src/main/java/io/delta/dynamodbcommitstore/DynamoDBCommitOwnerClient.java:454]: Should we call this `writeActionsToBackfilledFile`?

File: spark/src/main/java/io/delta/dynamodbcommitstore/DynamoDBCommitOwnerClient.java [hunk 7]
```diff
@@ -0,0 +1,681 @@
...
+            Option<Object> lastKnownBackfilledVersion) {
+        GetCommitsResponse resp =
+                getCommits(logPath, managedCommitTableConf, lastKnownBackfilledVersion, Option.empty());
+        validateBackfilledFileExists(logPath, hadoopConf, lastKnownBackfilledVersion);
+        if (version > resp.getLatestTableVersion()) {
+            throw new IllegalArgumentException(
+                    "The requested backfill version " + version + " is greater than the latest " +
+                            "version " + resp.getLatestTableVersion() + " for the table.");
+        }
+        // If partial writes a visible in this filesystem, we should not try to overwrite existing
```
Comment (LukasRupprecht) [inline spark/src/main/java/io/delta/dynamodbcommitstore/DynamoDBCommitOwnerClient.java:510]:
```suggestion
        // If partial writes are visible in this filesystem, we should not try to overwrite existing
```

File: spark/src/main/scala/org/apache/spark/sql/delta/managedcommit/CommitOwnerClient.scala [hunk 8]
```diff
@@ -43,7 +43,7 @@ case class Commit(
 case class CommitFailedException(
     private val retryable: Boolean,
     private val conflict: Boolean,
-    private val message: String) extends Exception(message) {
+    private val message: String) extends RuntimeException(message) {
```
Comment (LukasRupprecht) [inline spark/src/main/scala/org/apache/spark/sql/delta/managedcommit/CommitOwnerClient.scala:46]: Why is this changed needed? Isn't that just making it more restrictive?

Comment (dhruvarya-db) [reply spark/src/main/scala/org/apache/spark/sql/delta/managedcommit/CommitOwnerClient.scala:49]: This is needed because the CommitOwnerClient trait does not declare that `commit` can throw `CommitFailedException` in the signature (Scala doesn't let us do that). The Java code won't compile if I try to throw an exception of type Exception if it has not been declared in the function signature. `RuntimeException`s can be thrown though.

**Review by prakharjain09 — APPROVED** [review]
